ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/PackageCommitter.scala

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils
+
+import java.util.concurrent.Executors
+import com.codahale.metrics.{Counter, Timer}
+import com.daml.ledger.participant.state.kvutils.Conversions.buildTimestamp
+import com.daml.ledger.participant.state.kvutils.DamlKvutils._
+import com.daml.ledger.participant.state.kvutils.committer.{Committer, Context}
+import com.digitalasset.daml.lf.archive.Decode
+import com.digitalasset.daml.lf.data.Ref
+import com.digitalasset.daml.lf.engine.Engine
+import com.digitalasset.daml.lf.language.Ast
+import com.digitalasset.daml_lf.DamlLf.Archive
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext
+
+private[kvutils] case class PackageCommitter(engine: Engine)
+    extends Committer[DamlPackageUploadEntry, DamlPackageUploadEntry.Builder] {
+
+  private val serialContext =
+    ExecutionContext.fromExecutor(Executors.newSingleThreadExecutor())
+
+  private object Metrics {
+    val preloadTimer: Timer = metricsRegistry.timer(metricsName("preload-timer"))
+    val decodeTimer: Timer = metricsRegistry.timer(metricsName("decode-timer"))
+    val accepts: Counter = metricsRegistry.counter(metricsName("accepts"))
+  }
+
+  override def init(subm: DamlPackageUploadEntry): DamlPackageUploadEntry.Builder =
+    subm.toBuilder
+
+  override val steps: Iterable[Step] = Iterable(
+    filterDuplicates,
+    (_, subm) => {
+      serialContext.execute(preload(subm.getArchivesList.asScala))
```

Comment:
I'm guessing we are okay with launching the preloading as a "hidden" side-effect. Maybe make this conscious decision explicit with a comment.

Reply:
We're not that ok with this approach as we'd much rather want to fully validate the packages before accepting them. This however is currently rather problematic with couple integrations as it's very slow. I'll add comments on this.